File: src/packets/PacketC2SMute.js
const PacketC2S = require("./PacketC2S");

/**
 * Client -> Server
 * ID = 0x0D
 * Mute
 * Data:
 *      (Unsigned Char) Mute state
 *      (Unsigned Short) Fade Time
 *      (Optional) (Unsigned Char) Controller ID
 *      (Optional) (Unsigned Char) Zone ID
 */
class PacketC2SMute extends PacketC2S {
    parseData() {
        this._muteState = this._buffer.readUInt8();
        this._fadeTime = this._buffer.readUInt16LE();
        this._controllerID = false;
        this._zoneID = false;

        if (this._buffer.remaining() > 1) {
            this._controllerID = this._buffer.readUInt8();
            this._zoneID = this._buffer.readUInt8();
        }
    }

    getMuteState() {
        return this._muteState;
    }

    getFadeTime() {
        return this._fadeTime;
    }

    getControllerID() {
        return this._controllerID;
    }

    getZoneID() {
        return this._zoneID;
    }

    getID() {
        return PacketC2SMute.ID;
    }
}

PacketC2SMute.ID = 0x0D;
PacketC2SMute.MUTE_OFF = 0x00;
PacketC2SMute.MUTE_ON = 0x01;
PacketC2SMute.MUTE_TOGGLE = 0x02;

module.exports = PacketC2SMute;

File: src/RNetPi.Core/Packets/PacketC2SMute.cs
namespace RNetPi.Core.Packets;

/// <summary>
/// Client -> Server
/// ID = 0x0D
/// Mute
/// Data:
///     (Unsigned Char) Mute state
///     (Unsigned Short) Fade Time
///     (Optional) (Unsigned Char) Controller ID
///     (Optional) (Unsigned Char) Zone ID
/// </summary>
public class PacketC2SMute : PacketC2S
{
    public const byte ID = 0x0D;
    
    public const byte MuteOff = 0x00;
    public const byte MuteOn = 0x01;
    public const byte MuteToggle = 0x02;

    public byte MuteState { get; private set; }
    public ushort FadeTime { get; private set; }
    public byte? ControllerID { get; private set; }
    public byte? ZoneID { get; private set; }

    public PacketC2SMute(byte[] data) : base(data)
    {
    }

    public override byte GetID() => ID;

    protected override void ParseData()
    {
        MuteState = Reader.ReadByte();
        FadeTime = Reader.ReadUInt16();
        
        // Check if there's remaining data for controller and zone ID
        if (Reader.BaseStream.Length - Reader.BaseStream.Position > 1)
        {
            ControllerID = Reader.ReadByte();
            ZoneID = Reader.ReadByte();
        }
    }
}